modules/core/src/v2/coins/algo.ts

```diff
@@ -1,35 +1,115 @@
 import BaseCoin = require('../baseCoin');
+const algosdk = require('algosdk');
+
+interface KeyPair {
+  pub: string;
+  prv: string;
+  secret: string;
+}
 
 class Algo extends BaseCoin {
+  constructor() {
+    super();
+  }
 
-    constructor() {
-      super();
-    }
+  getChain(): string {
+    return 'algo';
+  }
 
-    getChain() {
-        return 'algo';
-    }
+  getFamily(): string {
+    return 'algo';
+  }
 
-    getFamily() {
-        return 'algo';
-    }
+  getFullName(): string {
+    return 'Algorand';
+  }
 
-    getFullName() {
-        return 'Algorand';
-    }
-
-    getBaseFactor() {
-        return 1e6;
-    }
+  getBaseFactor(): number {
+    return 1e6;
+  }
 
-    /**
+  /**
      * Flag for sending value of 0
      * @returns {boolean} True if okay to send 0 value, false otherwise
      */
-    valuelessTransferAllowed(): boolean {
-        // TODO: this sounds like its true with the staking txes - confirm before launch
-        return false;
+  valuelessTransferAllowed(): boolean {
+    // TODO: this sounds like its true with the staking txes - confirm before launch
+    return false;
+  }
+
+  /**
+   * Generate ed25519 key pair
+   *
+   * @param seed
+   * @returns {Object} object with generated pub, prv and secret
+   */
+  generateKeyPair(seed?: Buffer): KeyPair {
+    const pair = seed ? algosdk.generateAccountFromSeed(seed) : algosdk.generateAccount();
+    return {
+      pub: pair.addr, // encoded pub
+      prv: algosdk.Seed.encode(pair.sk), // encoded seed
+      secret: pair.sk, // secret as 64 byte long Uint8Array
+    };
+  }
+
+  /**
+   * Return boolean indicating whether input is valid public key for the coin.
+   *
+   * @param {String} pub the pub to be checked
+   * @returns {Boolean} is it valid?
+   */
+  isValidPub(pub): boolean {
+    return algosdk.isValidAddress(pub);
+  }
+
+  /**
+   * Return boolean indicating whether input is valid seed for the coin
+   * In Algorand, when the private key is encoded as base32 string only the first 32 bytes are taken,
+   * so the encoded value is actually the seed
+   *
+   * @param {String} prv the prv to be checked
+   * @returns {Boolean} is it valid?
+   */
+  isValidPrv(prv): boolean {
+    return algosdk.isValidSeed(prv);
+  }
+
+  /**
+   * Return boolean indicating whether input is valid public key for the coin
+   *
+   * @param {String} address the pub to be checked
+   * @returns {Boolean} is it valid?
+   */
+  isValidAddress(address): boolean {
+    return algosdk.isValidAddress(address);
+  }
+
+  /**
+   * Sign message with private key
+   *
+   * @param key
+   * @param message
+   */
+  signMessage(key, message): Buffer {
+    let seed = key.prv;
+    if (!this.isValidPrv(seed)) {
+      throw new Error(`invalid seed: ${seed}`);
     }
+    if (typeof seed === 'string') {
+      try {
+        seed = algosdk.Seed.decode(seed).seed;
+      } catch (e) {
+        throw new Error(`invalid seed: ${seed}`);
```

Comment:
Please throw with different message, so we can disambiguate from the first case when we see this in a log from a customer.